### PR TITLE
add support for AAD Pfx Cert Auth

### PR DESCRIPTION
**Pull request by abmo-x**
#### Pull Request Description
Support Pfx certificate based AAD App Authentication
---

#### Future Release Comment
Support Pfx certificate based AAD App Authentication

**Breaking Changes:**
- None

**Features:**
- Pfx cert auth support

**Fixes:**
- None

### CODE REVIEW

File: connector/src/main/scala/com/microsoft/kusto/spark/utils/KustoClientCache.scala [hunk 1]
```diff
@@ -26,6 +26,12 @@ object KustoClientCache {
         ConnectionStringBuilder.createWithAadApplicationCredentials(aliasAndAuth.engineUri, app.ID, app.password, app.authority),
         ConnectionStringBuilder.createWithAadApplicationCredentials(aliasAndAuth.ingestUri, app.ID, app.password, app.authority)
       )
+      case app: AadApplicationCertificateAuthentication =>
+        val keyCert = CertUtils.readPfx(app.certFilePath, app.certPassword)
+        (
+          ConnectionStringBuilder.createWithAadApplicationCertificate(aliasAndAuth.engineUri, app.ID, keyCert.cert, keyCert.key),
+          ConnectionStringBuilder.createWithAadApplicationCertificate(aliasAndAuth.engineUri, app.ID, keyCert.cert, keyCert.key)
```
Comment (ohadbitt) [inline connector/src/main/scala/com/microsoft/kusto/spark/utils/KustoClientCache.scala:33]:
```suggestion
          ConnectionStringBuilder.createWithAadApplicationCertificate(aliasAndAuth.ingestUri, app.ID, keyCert.cert, keyCert.key)
```

File: pom.xml [hunk 2]
```diff
@@ -9,7 +9,7 @@
     <version>${revision}</version>
 
     <properties>
-        <revision>2.0.1</revision>
+        <revision>2.0.2</revision>
```
Comment (ohadbitt) [inline pom.xml:12]: 2.1.0 
we update small version on bug fix , second on new features

File: connector/pom.xml [hunk 3]
```diff
@@ -218,6 +218,7 @@
                     <excludes>
                         <exclude>**/*E2E.scala</exclude>
                     </excludes>
+                    <argLine>-Xms512m -Xmx512m</argLine>
```
[the remark this answers is not in the document]
Comment (abmo-x) [reply connector/pom.xml:221]: ya, spark session fails to start due to low default memory:
 Cause: java.lang.IllegalArgumentException: System memory 259522560 must be at least 471859200. Please increase heap size using the --driver-memory option or spark.driver.memory in Spark configuration.
  at org.apache.spark.memory.UnifiedMemoryManager$.getMaxMemory(UnifiedMemoryManager.scala:217)
  at org.apache.spark.memory.UnifiedMemoryManager$.apply(UnifiedMemoryManager.scala:199)
  at org.apache.spark.SparkEnv$.create(SparkEnv.scala:330)
  at org.apache.spark.SparkEnv$.createDriverEnv(SparkEnv.scala:175)
  at org.apache.spark.SparkContext.createSparkEnv(SparkContext.scala:257)
  at org.apache.spark.SparkContext.<init>(SparkContext.scala:424)


I can set just -Xms flag, that should start the jvm with enough minimum memory.

File: connector/src/main/scala/com/microsoft/kusto/spark/utils/CertUtils.scala [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+      // Iterate over all aliases to find the private key
+      val aliases = store.aliases
+      var alias = ""
+      // Break if alias refers to a private key because we want to use that
+      // certificate
+      while ( aliases.hasMoreElements && !isAliasWithPrivateKey) {
+        alias = aliases.nextElement
+        isAliasWithPrivateKey = store.isKeyEntry(alias)
+      }
+      if (isAliasWithPrivateKey) { // Retrieves the certificate from the Java keystore
```
[the remark this answers is not in the document]
Comment (abmo-x) [reply connector/src/main/scala/com/microsoft/kusto/spark/utils/CertUtils.scala]: done

File: connector/src/main/scala/com/microsoft/kusto/spark/authentication/KustoAuthentication.scala [hunk 5]
```diff
@@ -20,6 +20,17 @@ case class AadApplicationAuthentication(ID: String, password: String, authority:
   override def hashCode(): Int = ID.hashCode + authority.hashCode
 }
 
+case class AadApplicationCertificateAuthentication(ID: String, certFilePath: String, certPassword: String) extends KustoAuthentication {
```
[the remark this answers is not in the document]
Comment (abmo-x) [reply connector/src/main/scala/com/microsoft/kusto/spark/authentication/KustoAuthentication.scala]: done

File: connector/src/main/scala/com/microsoft/kusto/spark/authentication/KustoAuthentication.scala [hunk 6]
```diff
@@ -20,6 +20,17 @@ case class AadApplicationAuthentication(ID: String, password: String, authority:
...
+  def canEqual(that: Any) : Boolean = that.isInstanceOf[KustoAccessTokenAuthentication]
```
[the remark this answers is not in the document]
Comment (abmo-x) [reply connector/src/main/scala/com/microsoft/kusto/spark/authentication/KustoAuthentication.scala]: fixed

File: connector/src/main/scala/com/microsoft/kusto/spark/utils/KustoClientCache.scala (hunk 1 above)
Comment (abmo-x) [reply connector/src/main/scala/com/microsoft/kusto/spark/utils/KustoClientCache.scala]: fixed

File: pom.xml (hunk 2 above)
Comment (abmo-x) [reply pom.xml]: fixed

**Review by ohadbitt — APPROVED** [review]

**Pull request merged**